cpp/include/copying.hpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef COPYING_HPP
+#define COPYING_HPP
+
+#include "cudf.h"
+#include "types.hpp"
+
+namespace cudf {
+/**
+ * @brief Scatters the rows (including null values) of a set of source columns
+ * into a set of destination columns.
+ *
+ * Scatters the rows of the source columns into the destination columns
+ * according to a scatter map such that row "i" from the source columns will be
+ * scattered to row "scatter_map[i]" in the destination columns.
+ *
+ * The datatypes between coresponding columns in the source and destination
+ * columns must be the same.
+ *
+ * The number of elements in the scatter_map must equal the number of rows in
+ * the source columns.
+ *
+ * If any index in scatter_map is outside the range of [0, num rows in
+ * destination_columns), the result is undefined.
+ *
+ * If the same index appears more than once in scatter_map, the result is
+ * undefined.
+ *
+ * @Param[in] source_table The columns whose rows will be scattered
+ * @Param[in] scatter_map An array that maps rows in the input columns
+ * to rows in the output columns.
+ * @Param[out] destination_table A preallocated set of columns with a number
+ * of rows equal in size to the maximum index contained in scatter_map
+ *
+ * @Returns GDF_SUCCESS upon successful completion
+ */
+gdf_error scatter(table const* source_table, gdf_index_type const scatter_map[],
```

Comment:
Where is the C API wrapper defined for scatter and gather?

Reply:
There isn't one yet.

We haven't defined a structure or procedure for defining C wrappers of C++ APIs. I see creating the C wrapper to be analogous to creating Cython wrappers, and we need a well-defined place to put them and what they're going to look like.

I'd be happy to prototype it in this PR.

Reply:
I assumed functions.h was the home of the C API (interface), as always. I think the implementation can just be an `extern "C"` function in the same location (even the same file if it makes sense) as the feature's C++ implementation.

Reply:
I think we'd like to move away from having a monolithic `functions.h`.

My thought would be to have something like a `wrappers` directory that has a `C` sub-directory, which then contains `feature1.h`, `feature1.h` that corresponds to `feature1.hpp` of the C++ API.

Reply:
I doubt you will get agreement from the Python side to move all of the Python API under a wrappers directory...

Reply:
The way Arrow and other cross-language libraries typically do it is have separate top-level folders for different languages. I.E. `cpp`, `c`, `python`, `R`, etc.

Reply:
> The way Arrow and other cross-language libraries typically do it is have separate top-level folders for different languages. I.E. `cpp`, `c`, `python`, `R`, etc.

My only concern with this solution is that it places the `c` and `cpp` directories on the same level when in reality, the `cpp` directory is where the real code lives and the `c` directory is just wrappers.

Reply:
I can see that, but I don't think it's a big issue.

Reply:
Based on external conversations, implementing a C API is considered future work to be done when and if necessary pending future design discussions.